### PR TITLE
Benchmarks for more lib.distances methods required #3205

### PR DESCRIPTION
**Pull request by manishsaini6421**
Fixes #

Changes made in this Pull Request:
 - I added benchmarks in lib.distances for the following functions .
    lib.distances.calc_bonds
    lib.distances.calc_angles
    lib.distances.calc_dihedrals

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/bat.py [hunk 1]
```diff
@@ -63,7 +63,7 @@
 --------
 :class:`~MDAnalysis.analysis.dihedrals.Dihedral`
    class to calculate dihedral angles for a given set of atoms or residues
-:func:`MDAnalysis.lib.distances.calc_dihedrals()`
+:func:`MDAnalysis.calc_dihedrals()`
```
Comment (IAlibay) [inline package/MDAnalysis/analysis/bat.py]: Why is this change made here? There is no such method as `MDAnalysis.calc_dihedrals()`, the previous name was fine?

File: package/MDAnalysis/lib/distances.py [hunk 2]
```diff
@@ -1242,6 +1243,7 @@ def transform_StoR(coords, box, backend="serial"):
 
 @check_coords('coords1', 'coords2')
 def calc_bonds(coords1, coords2, box=None, result=None, backend="serial"):
+    start = time.perf_counter()
```
Comment (IAlibay) [inline package/MDAnalysis/lib/distances.py]: This isn't how we do benchmarks in MDAnalysis, please see @hmacdope's comment here about how we use ASV for benchmarking https://github.com/MDAnalysis/mdanalysis/issues/3205#issuecomment-1079613028

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/analysis/bat.py (hunk 1 above)
Comment (manishsaini6421) [reply package/MDAnalysis/analysis/bat.py]: @IAlibay i don't know how it changed , i made changes only in [package/MDAnalysis/lib/distances.py](https://github.com/MDAnalysis/mdanalysis/blob/d9272baa41f8cdbff30869e3ddfaf1bec7e07a8f/package/MDAnalysis/lib/distances.py) . sorry for that i'll fix it again

File: package/MDAnalysis/lib/distances.py (hunk 2 above)
Comment (manishsaini6421) [reply package/MDAnalysis/lib/distances.py]: @IAlibay i'll do it according to https://github.com/MDAnalysis/mdanalysis/issues/3205#issuecomment-1079613028

Comment (manishsaini6421) [reply package/MDAnalysis/lib/distances.py]: @IAlibay After reading https://github.com/MDAnalysis/mdanalysis/issues/3205#issuecomment-1079613028 and related pages i did not understand that where to add banchmarks? ,which files i have to edit ? what should i write?
please help me in this.

> This isn't how we do benchmarks in MDAnalysis, please see @hmacdope's comment here about how we use ASV for benchmarking [#3205 (comment)](https://github.com/MDAnalysis/mdanalysis/issues/3205#issuecomment-1079613028)

Comment (manishsaini6421) [reply package/MDAnalysis/lib/distances.py]: @IAlibay Now I have added banchmarks according to https://github.com/MDAnalysis/mdanalysis/issues/3205#issuecomment-1079613028 .please review it again

**Comment by manishsaini6421** [comment]
@IAlibay please review it asap ,i have to submit final application by tomorrow.

**Comment by lilyminium** [comment]
@manishsaini6421 we've repeatedly asked you to please be patient. Be respectful of our free labor: It’s a public holiday weekend and we do this in our free time. Your request is not urgent and will be dealt with when possible.

For Outreachy, we are closing applications on Tuesday 19 April 5 pm UK time. You still have more than 30 hours. The warning about not accepting new PRs was for new PRs opened after that date. You also already have merged a contribution to MDAnalysis.

**Comment by manishsaini6421** [comment]
> @manishsaini6421 we've repeatedly asked you to please be patient. Be respectful of our free labor: It’s a public holiday weekend and we do this in our free time. Your request is not urgent and will be dealt with when possible.
> 
> For Outreachy, we are closing applications on Tuesday 19 April 5 pm UK time. You still have more than 30 hours. The warning about not accepting new PRs was for new PRs opened after that date. You also already have merged a contribution to MDAnalysis.

@lilyminium Thanks for clearing that up, i was so worried about it.

**Review by richardjgowers — DISMISSED** [review]
Thanks for your contribution.

The `ag_methods.py` file is for atomgroup methods.  There's another file in `benchmarks/` that handles `lib.distances` code and these benchmarks should be moved there.

To property benchmark code you need some representative usage of the code, rather than just calling the function with one angle/dihedral.  Typically people will calculate 1,000s of angles at once,.  Benchmarking the calculation of a single angle is just timing the function overheads.

Once you find the correct file to place these benchmarks you'll see how to generate enough random data to properly benchmark the functions here.

**Comment by manishsaini6421** [comment]
@richardjgowers I have moved new created banchmarks to other file and also updated according to other banchmark functions

**Comment by manishsaini6421** [comment]
@richardjgowers sir please review it ,so i can work on another issue.

File: benchmarks/benchmarks/ag_methods.py [hunk 3]
```diff
@@ -200,7 +201,7 @@ def time_wrap(self, num_atoms):
         atomgroup with default params.
         """
         self.ag.wrap()
-
+
```
Comment (hmacdope) [inline benchmarks/benchmarks/ag_methods.py]: Blank line with whitespace not required.

File: benchmarks/benchmarks/ag_methods.py [hunk 4]
```diff
@@ -32,6 +32,7 @@ def setup(self, num_atoms):
         self.rot_matrix = np.ones((3,3))
         self.trans = np.ones((4,4))
 
+
```
Comment (hmacdope) [inline benchmarks/benchmarks/ag_methods.py]: Is this blank line change needed?

File: benchmarks/benchmarks/ag_methods.py [hunk 5]
```diff
@@ -413,3 +414,7 @@ def setup(self, universe_type):
 
     def time_find_cached_fragments(self, universe_type):
         frags = self.u.atoms.fragments
+
```
Comment (hmacdope) [inline benchmarks/benchmarks/ag_methods.py]: Again more blank line changes not required.

File: benchmarks/benchmarks/analysis/distances.py [hunk 6]
```diff
@@ -128,3 +132,32 @@ def time_between(self, num_atoms):
...
                           B=self.ag2,
                           distance=15.0)
+
+    def time_calc_bonds(self,num_atoms):
+        np.random.seed(17809)
+        self.coords_1 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+        np.random.seed(9008716)
+        self.coords_2 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+       
+        MDAnalysis.lib.distances.calc_bonds(self.coords_1,self.coords_2)
```
Comment (hmacdope) [inline benchmarks/benchmarks/analysis/distances.py]: Can you change the import in this file to the idiomatic `import MDAnalysis as mda` and make the corresponding changes to the function calls?

File: benchmarks/benchmarks/analysis/distances.py [hunk 7]
```diff
@@ -128,3 +132,32 @@ def time_between(self, num_atoms):
...
+
+    def time_calc_angles(self,num_atoms):
+        np.random.seed(17809)
```
Comment (hmacdope) [inline benchmarks/benchmarks/analysis/distances.py]: These are already set in `setup()` you don't need to remake the arrays in each benchmark. In fact the way you have  it here the **array creation** is also being benchmarked.

File: benchmarks/benchmarks/analysis/distances.py [hunk 8]
```diff
@@ -128,3 +132,32 @@ def time_between(self, num_atoms):
...
+        self.coords_1 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+        np.random.seed(9008716)
+        self.coords_2 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+        np.random.seed(15809)
+        self.coords_3 = np.random.random_sample((num_atoms, 3)).astype(np.float32)
+     
+        MDAnalysis.lib.distances.calc_angles(self.coords_1,self.coords_2,self.coords_3)
+
+    def time_calc_dihedrals(self,num_atoms):
+        np.random.seed(17809)
```
Comment (hmacdope) [inline benchmarks/benchmarks/analysis/distances.py]: Same here with the arrays.

**Review by hmacdope — CHANGES_REQUESTED** [review]
A decent start. See my comments and make sure that you are only benchmarking the function itself and not associated setup costs.

**Comment by manishsaini6421** [comment]
@hmacdope done changes that were required.

File: benchmarks/benchmarks/ag_methods.py (hunk 4 above)
Comment (IAlibay) [reply benchmarks/benchmarks/ag_methods.py]: As above - this still needs fixing.

File: benchmarks/benchmarks/ag_methods.py [hunk 9]
```diff
@@ -199,8 +199,7 @@ def time_wrap(self, num_atoms):
         """Benchmark wrap() operation on
         atomgroup with default params.
         """
-        self.ag.wrap()
-
+        self.ag.wrap()
```
Comment (IAlibay) [inline benchmarks/benchmarks/ag_methods.py]: Please remove the extra blank space at the end of this line.

File: benchmarks/benchmarks/ag_methods.py [hunk 10]
```diff
@@ -412,4 +411,4 @@ def setup(self, universe_type):
         frags = self.u.atoms.fragments  # Priming the cache
 
     def time_find_cached_fragments(self, universe_type):
-        frags = self.u.atoms.fragments
+        frags = self.u.atoms.fragments
```
Comment (IAlibay) [inline benchmarks/benchmarks/ag_methods.py]: Please remove the extra blank spaces introduced here at the end of the line.

File: benchmarks/benchmarks/analysis/distances.py [hunk 11]
```diff
@@ -128,3 +133,12 @@ def time_between(self, num_atoms):
                           A=self.ag1,
                           B=self.ag2,
                           distance=15.0)
+
+    def time_calc_bonds(self,num_atoms):
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]: blank space between the two arguments

File: benchmarks/benchmarks/analysis/distances.py [hunk 12]
```diff
@@ -128,3 +133,12 @@ def time_between(self, num_atoms):
...
+        mda.lib.distances.calc_bonds(self.coords_1,self.coords_2)
+
+    def time_calc_angles(self,num_atoms):
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]: As above

File: benchmarks/benchmarks/analysis/distances.py [hunk 13]
```diff
@@ -128,3 +133,12 @@ def time_between(self, num_atoms):
...
+        mda.lib.distances.calc_angles(self.coords_1,self.coords_2,self.coords_3)
+
+    def time_calc_dihedrals(self,num_atoms):
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]: As above

File: benchmarks/benchmarks/analysis/distances.py [hunk 14]
```diff
@@ -128,3 +133,12 @@ def time_between(self, num_atoms):
                           A=self.ag1,
                           B=self.ag2,
                           distance=15.0)
+
+    def time_calc_bonds(self,num_atoms):
+        mda.lib.distances.calc_bonds(self.coords_1,self.coords_2)
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]: blank space between arguments to calc_bonds

File: benchmarks/benchmarks/analysis/distances.py [hunk 15]
```diff
@@ -128,3 +133,12 @@ def time_between(self, num_atoms):
...
+
+    def time_calc_angles(self,num_atoms):
+        mda.lib.distances.calc_angles(self.coords_1,self.coords_2,self.coords_3)
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]: blank space between arguments to calc_angles

File: benchmarks/benchmarks/analysis/distances.py [hunk 16]
```diff
@@ -128,3 +133,12 @@ def time_between(self, num_atoms):
...
+
+    def time_calc_dihedrals(self,num_atoms):
+        mda.lib.distances.calc_dihedrals(self.coords_1,self.coords_2,self.coords_3,self.coords_4)
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]: blank space between arguments to calc_dihedrals

**Review by IAlibay — DISMISSED** [review]
I'll let @hmacdope re-review, but overall please do review formatting of your code for any PEP8 violations.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Can you add a `CHANGELOG` entry and fix the last of the PEP8 issues please :)

File: package/CHANGELOG [hunk 17]
```diff
@@ -61,6 +61,10 @@ Enhancements
     analysis.nucleicacids
   * CRD extended format can now be explicitly requested with the `extended`
     keyword (Issue #3605)
+  * Added benchmarks in lib.distances for the following functions .
```
Comment (hmacdope) [inline package/CHANGELOG]: Can you make this a fewer lines? Other than that LGTM.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Last thing and then good to go.

File: package/CHANGELOG (hunk 17 above)
Comment (manishsaini6421) [reply package/CHANGELOG]: @hmacdope i am working on it.

**Comment by IAlibay** [comment]
Assigning this to you @hmacdope

**Review by hmacdope — CHANGES_REQUESTED** [review]
Can you fix the PEP8 issues?

**Comment by hmacdope** [comment]
See the comment by PEP8 bot for the things that need to change.

**Comment by manishsaini6421** [comment]
> See the comment by PEP8 bot for the things that need to change.

@hmacdope can you tell me, where the problem ? i am unable to find it.

**Comment by hmacdope** [comment]
> Hello @manishsaini6421! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`benchmarks/benchmarks/ag_methods.py`](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/ag_methods.py):
> 
> > [Line 415:39](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/ag_methods.py#L415): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file
> 
> * In the file [`benchmarks/benchmarks/analysis/distances.py`](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py):
> 
> > [Line 31:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L31): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> > [Line 33:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L33): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> > [Line 35:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L35): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> > [Line 37:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L37): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> > [Line 40:42](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L40): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> > [Line 130:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L130): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> > [Line 142:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> 
> ##### Comment last updated at 2022-05-16 04:20:51 UTC
See this comment @manishsaini6421

**Comment by manishsaini6421** [comment]
> > Hello @manishsaini6421! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> > 
> > * In the file [`benchmarks/benchmarks/ag_methods.py`](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/ag_methods.py):
> > 
> > > [Line 415:39](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/ag_methods.py#L415): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file
> > 
> > 
> > 
> > * In the file [`benchmarks/benchmarks/analysis/distances.py`](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py):
> > 
> > > [Line 31:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L31): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> > > [Line 33:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L33): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> > > [Line 35:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L35): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> > > [Line 37:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L37): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> > > [Line 40:42](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L40): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> > > [Line 130:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L130): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> > > [Line 142:80](https://github.com/MDAnalysis/mdanalysis/blob/de263e84a46ade70bf8a826e72da6c2324f46ca1/benchmarks/benchmarks/analysis/distances.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> > 
> > 
> > ##### Comment last updated at 2022-05-16 04:20:51 UTC
> 
> See this comment @manishsaini6421

@hmacdope i Fixed it, please review it again.

**Comment by hmacdope** [comment]
@manishsaini6421, the PEP8 bot still has a few things to fix. Please address them. :)

File: benchmarks/benchmarks/ag_methods.py [hunk 18]
```diff
@@ -413,3 +413,4 @@ def setup(self, universe_type):
 
     def time_find_cached_fragments(self, universe_type):
         frags = self.u.atoms.fragments
+
```
Comment (IAlibay) [inline benchmarks/benchmarks/ag_methods.py]:
```suggestion
```

File: benchmarks/benchmarks/analysis/distances.py [hunk 19]
```diff
@@ -19,112 +20,147 @@ class DistancesBench(object):
...
+            (num_atoms, num_atoms), dtype=np.float64
+        )
+        self.array_shape_1D = int(num_atoms * (num_atoms - 1) / 2.0)
+        self.allocated_array_1D = np.empty(
+            self.array_shape_1D, dtype=np.float64
+        )
+        self.u = mda.Universe(GRO)
         self.ag1 = self.u.atoms[:num_atoms]
-        self.ag2 = self.u.atoms[num_atoms: 2 * num_atoms]
+        self.ag2 = self.u.atoms[num_atoms : 2 * num_atoms]
```
Comment (IAlibay) [inline benchmarks/benchmarks/analysis/distances.py]:
```suggestion
        self.ag2 = self.u.atoms[num_atoms: 2 * num_atoms]
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
@manishsaini6421 here are the changes that pep8speaks requires

File: benchmarks/benchmarks/analysis/distances.py (hunk 19 above)
Comment (manishsaini6421) [reply benchmarks/benchmarks/analysis/distances.py]: @IAlibay Fixed now, please review it again.

**Review by IAlibay — APPROVED** [review]
Given that I'm not super acquainted with the benchmarks stuff, I will only approve in the limit that the code quality seems reasonable (i.e. PEP8 wise), I'll let @hmacdope make the final call.

**Comment by manishsaini6421** [comment]
> Given that I'm not super acquainted with the benchmarks stuff, I will only approve in the limit that the code quality seems reasonable (i.e. PEP8 wise), I'll let @hmacdope make the final call.

@IAlibay ok thanks.

**Review by hmacdope — APPROVED** [review]
Excellent work @manishsaini6421. Thanks for your contribution! It is very much appreciated.

**Comment by hmacdope** [comment]
@richardjgowers do you have any blocking changes left to address? Otherwise I will merge.

**Comment by manishsaini6421** [comment]
> Excellent work @manishsaini6421. Thanks for your contribution! It is very much appreciated.

@hmacdope Thanks for your support.

**Comment by hmacdope** [comment]
Ping @richardjgowers as blocking :)

**Pull request merged**